language/move-lang/src/lib.rs

```diff
@@ -197,6 +210,88 @@ pub fn output_compiled_units(
     Ok(())
 }
 
+fn generate_interface_files_for_deps(
+    deps: &mut Vec<String>,
+    interface_file_output_directory_opt: Option<String>,
+) -> anyhow::Result<()> {
+    if let Some(dir) = generate_interface_files(deps, interface_file_output_directory_opt)? {
+        deps.push(dir)
+    }
+    Ok(())
+}
+
+pub fn generate_interface_files(
+    mv_file_locations: &[String],
+    interface_file_output_directory_opt: Option<String>,
+) -> anyhow::Result<Option<String>> {
+    let mv_files = find_move_filenames_by_extension(mv_file_locations, MOVE_COMPILED_EXTENSION)?
+        .into_iter()
+        .filter(|s| {
+            let p = Path::new(s);
+            let mv_file = p.is_file()
+                && p.extension()
+                    .map(|ext| ext == MOVE_COMPILED_EXTENSION)
+                    .unwrap_or(false);
+            mv_file || has_compiled_module_magic_number(s)
+        })
+        .collect::<Vec<_>>();
+    if mv_files.is_empty() {
+        return Ok(None);
+    }
+
+    let root_dir = interface_file_output_directory_opt
+        .unwrap_or_else(|| command_line::DEFAULT_OUTPUT_DIR.to_string());
+    let hash_dir = {
+        use std::{
+            collections::hash_map::DefaultHasher,
+            hash::{Hash, Hasher},
+        };
+
+        let mut hasher = DefaultHasher::new();
+        for mv_file in &mv_files {
+            std::fs::read(mv_file)?.hash(&mut hasher)
+        }
+        // TODO might want a better temp file setup here
+        let dir = format!(
+            "{}/{}_interfaces/{:020}",
+            root_dir,
+            MOVE_COMPILED_EXTENSION,
+            hasher.finish(),
+        );
+
+        dir
+    };
+
+    for mv_file in mv_files {
+        let (id, interface_contents) = interface_generator::write_to_string(&mv_file)?;
+        let addr_dir = format!("{}/{:#x}", hash_dir, id.address());
+        let file_path = format!("{}/{}.{}", addr_dir, id.name(), MOVE_EXTENSION);
+        // it's possible some files exist but not others due to multithreaded environments
+        if Path::new(&file_path).is_file() {
```

Comment:
This hashing story is a bit weird. 
Breakdown of what's going on
- I tried using temp files, but I had issues trying to convince the system not to clean them up before they were used
- I tried just using the timestamp or something to just make a new folder every time, but this generated way too many files for test runs
- The hashing was an attempt to just use one set of files per set of dependencies 
- But by grouping everything together, I had issues when the files had not yet been created. So each test thread would be racing to complete them, and then stepping on each other to write to the file. 
- Tried just checking to see if the directory existed, but that was too fast, as the thread might move on before all the files were populated by a different test thread 
- I tried doing some big hacks by having a `static MUTEX: Lazy<Mutex<()>> = ...` and then wrapping this whole section in a mutex, but I don't think it was very clean, and seemed to cause the tests to be really slow
- My solution was to just skip any file that was already created, and then the threads sort of helped each other in filling in the directory. It feels a bit unsafe as there still is technically a possibility that one thread starts creating the files while the other ones race ahead to use them, but get there before the creation is finished. That hasn't happened yet in my local testing yet. 

Any thoughts? cc @vgao1996

Reply:
My gut feeling is that there must be a better way, but I haven't come up with anything in particular to suggest yet. This might be something that needs to evolve over time.

Reply:
I've done the following as a better current solution

- Check if the file is there, if it is continue
- Write the file contents to a temp file
- Double check if the file is there, if it is continue
- std::fs::rename the temp file into the proper file. This should stop issues I had with one thread writing to the file as the other is reading from it